messagebus/src/vespa/messagebus/network/rpctarget.cpp

```diff
@@ -25,11 +25,12 @@ RPCTarget::~RPCTarget()
 void
 RPCTarget::resolveVersion(duration timeout, RPCTarget::IVersionHandler &handler)
 {
-    bool hasVersion = false;
     bool shouldInvoke = false;
-    {
+    ResolveState state = _state.load(std::memory_order_relaxed);
+    bool hasVersion = (state == VERSION_RESOLVED);
+    if ( ! hasVersion ) {
         vespalib::MonitorGuard guard(_lock);
-        if (_state == VERSION_RESOLVED || _state == PROCESSING_HANDLERS) {
+        if (state == PROCESSING_HANDLERS) {
```

Comment:
By removing the check for `_state == VERSION_RESOLVED`, doesn't this open up for the following race condition?
1. Thread reads `_state` == `TARGET_INVOKED` cached into local `state` variable, attempts to enter critical section. Is put on hold by kernel.
1. MBus thread receives resolved version, enters critical section, finishes processing handlers and finally sets `_state = VERSION_RESOLVED`
1. Original thread wakes up, enters critical section
1. Original thread sees that cached `state` value is _not_ equal to `PROCESSING_HANDLERS`, does not enter branch that can set `hasVersion = true`
1. Original thread initiates new version fetch towards peer even though version has already been resolved
1. New round of same race condition possible by any other subsequent concurrent senders

Note that it should be safe to use the original expression with both loads being made explicitly relaxed, as visibility should be guaranteed from entering the critical section. That should avoid the race condition while not introducing any memory fences.

Reply:
Very good catch - Classic error.